File: 第二轮灰测/播放器dsv4p_ga_max_stantard_imdoing/tools/serve.js
#!/usr/bin/env node
/**
 * tools/serve.js — 零依赖本地静态服务器（Node 14+ 即可，Linux/Windows/macOS 通用）
 *
 * 为什么需要它：浏览器在 file:// 下会把同目录的视频当作「跨源资源」，
 * WebGL 就无法把视频帧上传成纹理。用 http:// 打开就完全没有这个限制。
 * 本服务器实现了 HTTP Range（206），这是视频拖动定位必须的。
 *
 * 用法：
 *   node tools/serve.js                 # http://127.0.0.1:8321
 *   node tools/serve.js --port 9000 --open
 *   node tools/serve.js --host 0.0.0.0  # 允许局域网访问（谨慎）
 */
'use strict';
const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');
const { spawn } = require('child_process');

const argv = process.argv.slice(2);
function arg(name, dflt) {
  const i = argv.indexOf('--' + name);
  return i >= 0 && argv[i + 1] && !argv[i + 1].startsWith('--') ? argv[i + 1] : dflt;
}
const ROOT = path.resolve(arg('root', path.join(__dirname, '..')));
const HOST = arg('host', '127.0.0.1');
let PORT = parseInt(arg('port', '8321'), 10);
const OPEN = argv.includes('--open');
const QUIET = argv.includes('--quiet');

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.htm': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.txt': 'text/plain; charset=utf-8',
  '.md': 'text/markdown; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.webp': 'image/webp',
  '.ico': 'image/x-icon',
  '.mp4': 'video/mp4',
  '.m4v': 'video/mp4',
  '.mov': 'video/quicktime',
  '.webm': 'video/webm',
  '.mkv': 'video/x-matroska',
  '.ogv': 'video/ogg',
  '.mp3': 'audio/mpeg',
  '.wav': 'audio/wav',
  '.zip': 'application/zip',
  '.ttf': 'font/ttf',
  '.woff2': 'font/woff2'
};

function safeJoin(root, reqPath) {
  const decoded = decodeURIComponent(reqPath.split('?')[0]);
  const p = path.normalize(path.join(root, decoded));
  if (!p.startsWith(root)) return null;   // 阻止 ../ 越权
  return p;
}

function send(res, code, headers, body) {
  res.writeHead(code, Object.assign({
    'Cache-Control': 'no-cache, no-store, must-revalidate',
    'X-Content-Type-Options': 'nosniff'
  }, headers));
  if (body) res.end(body); else res.end();
}

const server = http.createServer((req, res) => {
  const t0 = Date.now();
  let target = safeJoin(ROOT, url.parse(req.url).pathname || '/');
  if (!target) return send(res, 403, { 'Content-Type': 'text/plain; charset=utf-8' }, '403 越权路径');

  fs.stat(target, (err, st) => {
    if (!err && st.isDirectory()) {
      target = path.join(target, 'index.html');
      st = null;
      err = null;
      try { st = fs.statSync(target); } catch (e) { err = e; }
    }
    if (err || !st || !st.isFile()) {
      log(404, req, target, t0);
      return send(res, 404, { 'Content-Type': 'text/html; charset=utf-8' },
        '<h1>404</h1><p>找不到：' + req.url + '</p><p><a href="/">回到播放器</a></p>');
    }

    const ext = path.extname(target).toLowerCase();
    const type = MIME[ext] || 'application/octet-stream';
    const range = req.headers.range;

    // HTTP Range：视频精确定位必需
    if (range) {
      const m = /^bytes=(\d*)-(\d*)$/.exec(range.trim());
      if (m) {
        let start = m[1] === '' ? null : parseInt(m[1], 10);
        let end = m[2] === '' ? null : parseInt(m[2], 10);
        if (start === null && end !== null) { start = Math.max(0, st.size - end); end = st.size - 1; }
        if (start !== null && end === null) end = st.size - 1;
        if (start === null) { start = 0; end = st.size - 1; }
        if (start > end || start >= st.size) {
          return send(res, 416, { 'Content-Range': 'bytes */' + st.size });
        }
        end = Math.min(end, st.size - 1);
        res.writeHead(206, {
          'Content-Type': type,
          'Content-Length': end - start + 1,
          'Content-Range': 'bytes ' + start + '-' + end + '/' + st.size,
          'Accept-Ranges': 'bytes',
          'Cache-Control': 'no-cache'
        });
        fs.createReadStream(target, { start, end }).pipe(res);
        log(206, req, target, t0, (end - start + 1));
        return;
      }
    }

    res.writeHead(200, {
      'Content-Type': type,
      'Content-Length': st.size,
      'Accept-Ranges': 'bytes',
      'Cache-Control': 'no-cache, no-store, must-revalidate'
    });
    if (req.method === 'HEAD') return res.end();
    fs.createReadStream(target).pipe(res);
    log(200, req, target, t0, st.size);
  });
});

function log(code, req, file, t0, bytes) {
  if (QUIET) return;
  const rel = path.relative(ROOT, file) || '/';
  const size = bytes ? (bytes > 1048576 ? (bytes / 1048576).toFixed(1) + 'MB' : Math.round(bytes / 1024) + 'KB') : '-';
  console.log(`  ${code}  ${req.method}  ${rel}  ${size}  ${Date.now() - t0}ms`);
}

function openBrowser(u) {
  const cmd = process.platform === 'win32' ? 'cmd'
    : (process.platform === 'darwin' ? 'open' : 'xdg-open');
  const args = process.platform === 'win32' ? ['/c', 'start', '""', u] : [u];
  try {
    const p = spawn(cmd, args, { detached: true, stdio: 'ignore' });
    p.on('error', () => {});
    p.unref();
  } catch (e) { /* 打不开就算了，用户自己点链接 */ }
}

let tries = 0;
server.on('error', (err) => {
  if (err.code === 'EADDRINUSE' && tries < 12) {
    tries++;
    PORT++;
    console.log(`  端口被占用，换到 ${PORT} …`);
    server.listen(PORT, HOST);
  } else {
    console.error('启动失败：', err.message);
    process.exit(1);
  }
});

server.listen(PORT, HOST, () => {
  const u = `http://${HOST === '0.0.0.0' ? '127.0.0.1' : HOST}:${PORT}/index.html`;
  console.log('');
  console.log('  dsv4p max stantard imdoing — 本地服务器已启动');
  console.log('  根目录: ' + ROOT);
  console.log('  地址:   ' + u);
  console.log('  (Ctrl+C 退出)');
  console.log('');
  if (OPEN) openBrowser(u);
});

File: 第二轮灰测/播放器dsv4p_ga_max_stantard_imdoing/tools/serve.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
tools/serve.py — 零依赖本地静态服务器（Python 3.7+，Linux/Windows/macOS 通用）
与 tools/serve.js 等价，供没装 Node 的机器使用。实现了 HTTP Range（视频定位必需）。

用法:
    python3 tools/serve.py
    python3 tools/serve.py --port 9000 --open
"""
import argparse
import mimetypes
import os
import posixpath
import re
import socket
import sys
import threading
import urllib.parse
import webbrowser
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

ROOT = os.path.abspath(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))

EXTRA_MIME = {
    '.mp4': 'video/mp4', '.m4v': 'video/mp4', '.mov': 'video/quicktime',
    '.webm': 'video/webm', '.mkv': 'video/x-matroska', '.ogv': 'video/ogg',
    '.js': 'text/javascript; charset=utf-8', '.mjs': 'text/javascript; charset=utf-8',
    '.css': 'text/css; charset=utf-8', '.html': 'text/html; charset=utf-8',
    '.json': 'application/json; charset=utf-8', '.md': 'text/markdown; charset=utf-8',
    '.zip': 'application/zip',
}

RANGE_RE = re.compile(r'^bytes=(\d*)-(\d*)$')


def guess_type(path):
    ext = os.path.splitext(path)[1].lower()
    if ext in EXTRA_MIME:
        return EXTRA_MIME[ext]
    t, _ = mimetypes.guess_type(path)
    return t or 'application/octet-stream'


class Handler(BaseHTTPRequestHandler):
    server_version = 'dsv4p-serve/1.0'
    protocol_version = 'HTTP/1.1'
    quiet = False

    def log_message(self, fmt, *args):
        if not Handler.quiet:
            sys.stdout.write('  %s\n' % (fmt % args))

    def translate(self):
        raw = urllib.parse.urlparse(self.path).path
        raw = urllib.parse.unquote(raw)
        parts = [p for p in posixpath.normpath(raw).split('/') if p not in ('', '.', '..')]
        target = os.path.join(ROOT, *parts)
        target = os.path.abspath(target)
        if not target.startswith(ROOT):      # 阻止越权
            return None
        if os.path.isdir(target):
            target = os.path.join(target, 'index.html')
        return target

    def do_HEAD(self):
        self.serve(head_only=True)

    def do_GET(self):
        self.serve(head_only=False)

    def serve(self, head_only=False):
        target = self.translate()
        if not target or not os.path.isfile(target):
            body = ('<h1>404</h1><p>找不到：%s</p><p><a href="/">回到播放器</a></p>'
                    % urllib.parse.quote(self.path)).encode('utf-8')
            self.send_response(404)
            self.send_header('Content-Type', 'text/html; charset=utf-8')
            self.send_header('Content-Length', str(len(body)))
            self.end_headers()
            if not head_only:
                self.wfile.write(body)
            return

        size = os.path.getsize(target)
        ctype = guess_type(target)
        rng = self.headers.get('Range')
        start, end = 0, size - 1
        partial = False

        if rng:
            m = RANGE_RE.match(rng.strip())
            if m:
                g1, g2 = m.group(1), m.group(2)
                if g1 == '' and g2 != '':
                    start, end = max(0, size - int(g2)), size - 1
                elif g1 != '' and g2 == '':
                    start, end = int(g1), size - 1
                elif g1 != '' and g2 != '':
                    start, end = int(g1), min(int(g2), size - 1)
                if start > end or start >= size:
                    self.send_response(416)
                    self.send_header('Content-Range', 'bytes */%d' % size)
                    self.send_header('Content-Length', '0')
                    self.end_headers()
                    return
                partial = True

        length = end - start + 1
        self.send_response(206 if partial else 200)
        self.send_header('Content-Type', ctype)
        self.send_header('Content-Length', str(length))
        self.send_header('Accept-Ranges', 'bytes')
        self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
        if partial:
            self.send_header('Content-Range', 'bytes %d-%d/%d' % (start, end, size))
        self.end_headers()
        if head_only:
            return
        remaining = length
        with open(target, 'rb') as f:
            f.seek(start)
            while remaining > 0:
                chunk = f.read(min(262144, remaining))
                if not chunk:
                    break
                try:
                    self.wfile.write(chunk)
                except (BrokenPipeError, ConnectionResetError):
                    return
                remaining -= len(chunk)


def free_port(host, port, tries=12):
    for i in range(tries):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            s.bind((host, port + i))
            s.close()
            return port + i
        except OSError:
            s.close()
    return port


def main():
    ap = argparse.ArgumentParser(description='dsv4p 本地静态服务器（零依赖）')
    ap.add_argument('--port', type=int, default=8321)
    ap.add_argument('--host', default='127.0.0.1')
    ap.add_argument('--open', action='store_true')
    ap.add_argument('--quiet', action='store_true')
    args = ap.parse_args()
    Handler.quiet = args.quiet

    port = free_port(args.host, args.port)
    httpd = ThreadingHTTPServer((args.host, port), Handler)
    url = 'http://%s:%d/index.html' % ('127.0.0.1' if args.host == '0.0.0.0' else args.host, port)
    print('')
    print('  dsv4p max stantard imdoing — 本地服务器已启动')
    print('  根目录: %s' % ROOT)
    print('  地址:   %s' % url)
    print('  (Ctrl+C 退出)')
    print('')
    if args.open:
        threading.Timer(0.6, lambda: webbrowser.open(url)).start()
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print('\n  已退出。')


if __name__ == '__main__':
    main()
